cmd/clusterctl/client/cluster/crd_migration.go

```diff
@@ -0,0 +1,224 @@
+		obj := objs[i]
+
+		if obj.GetKind() == "CustomResourceDefinition" {
+			crd := &apiextensionsv1.CustomResourceDefinition{}
+			if err := scheme.Scheme.Convert(&obj, crd, nil); err != nil {
+				return errors.Wrapf(err, "failed to convert CRD %q", obj.GetName())
+			}
+
+			if _, err := m.run(ctx, crd); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// run migrates CRs of a new CRD.
+// This is necessary when the new CRD drops a version which
+// was previously used as a storage version.
+func (m *crdMigrator) run(ctx context.Context, newCRD *apiextensionsv1.CustomResourceDefinition) (bool, error) {
+	log := logf.Log
+
+	// Gets the list of version supported by the new CRD
+	newVersions := sets.NewString()
+	for _, version := range newCRD.Spec.Versions {
+		newVersions.Insert(version.Name)
+	}
+
+	// Get the current CRD.
+	currentCRD := &apiextensionsv1.CustomResourceDefinition{}
+	if err := retryWithExponentialBackoff(newReadBackoff(), func() error {
+		return m.Client.Get(ctx, client.ObjectKeyFromObject(newCRD), currentCRD)
+	}); err != nil {
+		// Return if the CRD doesn't exist yet. We only have to migrate if the CRD exists already.
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	// Get the storage version of the current CRD.
+	currentStorageVersion, err := storageVersionForCRD(currentCRD)
+	if err != nil {
+		return false, err
+	}
+
+	// Return an error, if the current storage version has been dropped in the new CRD.
+	if !newVersions.Has(currentStorageVersion) {
+		return false, errors.Errorf("unable to upgrade CRD %q because the new CRD does not contain the storage version %q of the current CRD, thus not allowing CR migration", newCRD.Name, currentStorageVersion)
+	}
+
+	currentStatusStoredVersions := sets.NewString(currentCRD.Status.StoredVersions...)
+
+	// If the new CRD still contains all current stored versions, nothing to do
+	// as no previous storage version will be dropped.
+	if newVersions.HasAll(currentStatusStoredVersions.List()...) {
+		log.V(2).Info("CRD migration check passed", "name", newCRD.Name)
+		return false, nil
+	}
+
+	// Otherwise a version that has been used as storage version will be dropped, so it is necessary to migrate all the
+	// objects and drop the storage version from the current CRD status before installing the new CRD.
+	// Ref https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects
+	// Note: We are simply migrating all CR objects independent of the version in which they are actually stored in etcd.
+	// This way we can make sure that all CR objects are now stored in the current storage version.
+	// Alternatively, we would have to figure out which objects are stored in which version but this information is not
+	// exposed by the apiserver.
+	storedVersionsToDelete := currentStatusStoredVersions.Difference(newVersions)
```

Comment:
storedVersionsToDelete is only for information purposes.

Ok for me, just want to make sure we did not initially want to skip if there are none to delete.

Reply:
I think we already do, if there are no versions to delete storedVersionsToDelete will be empty which is the same as:
```go
	if newVersions.HasAll(currentStatusStoredVersions.List()...) {
		log.V(2).Info("CRD migration check passed", "name", newCRD.Name)
		return false, nil
	}
```

Reply:
Ack, so in that case we will never even reach this line 👍 

As mentioned: ok for me to keep it for the users to see what was removed :-)